specification/dartLangSpec.tex

```diff
@@ -15945,23 +15951,46 @@ \subsection{Lexical Lookup}
   It is a compile-time error if $\ell$ does not have access to \THIS{}
   (\ref{classes}).
 \item
-  When $\ell$ has access to \THIS{}
-  it is a compile-time error if $\ell$ occurs in
-  a class \DefineSymbol{C} and
-  the interface of $C$ does not have a member named $n$;
-  and it is a compile-time error if $\ell$ occurs in
-  an extension \DefineSymbol{E},
-  and the interface of the \ON{} type of $E$ does not have a member named $n$.
-  \commentary{%
-    So it is an error if there is no member with the right basename at all,
-    and also if there is one such member,
-    but it is a non-setter and we are looking for a setter,
-    or vice versa.%
-  }
+  Consider the case where $\ell$ has access to \THIS,
+  and $\ell$ occurs in a class \DefineSymbol{C}.
+  A compile-time error occurs in each of the following two situations:
+  \begin{itemize}
+  \item
+    The interface of $C$ does not have a member named $n$,
```

Comment:
I don't know how to think about these checks here.  This feels very non-modular.  Why is it useful to catch some (but not all) of these kinds of errors here?  Why do we not just say that lexical lookup failed, and let the caller of lexical lookup figure out what to do?

Reply:
We would have to repeat the same rule about 30 times: It is an error if we find a getter but we're looking for a setter, and so on.

For instance, an identifier reference `id` would have that rule (at several locations in section `Identifier Reference`), and so would an unqualified  function invocation `id<...>(...)`, and an assignment `id = e`, and a composite assignment `id += e`, and so on.

It is true that we do not catch all errors: When there is no declaration with the requested basename in scope, and there is access to `this`, and the interface of the enclosing class does not have a member with the requested basename, we may still have multiple applicable extensions.

Do you think we have other missing errors? I think this one isn't so bad, because any reports about the error would deal with exactly the same problem even if we had had an explicit `this` in the first place. So it isn't going to be wrong if we simply proceed with `this.id...` and get the error later on.

Reply:
Ok, I think I might get what you're saying.  Paraphrasing, I think what you are saying is that there are really two different kinds of errors we might be concerned about:
  - We didn't find anything with the right name
     - This error cannot be handled here, because there are additional lookup steps that we need to follow if lexical lookup fails
  - We found something with the right name, but it was a setter and we wanted a getter or vice versa
     -   We can't treat this as just another case of "lexical lookup failed to find what we wanted" because we want to stop subsequent non-lexical lookup from happening.

Every caller of "lexical lookup" already has to deal with the first situation in which lexical lookup doesn't find what we want (which was why I suggested not checking things here) but what you're saying is that even when lexical lookup succeeds, we may still have an error, and it's better to catch that in one place, here.  

Is that a fair summary?

Reply:
I very nearly agree. ;-)

The approach I've taken will find the errors that we can characterize as lookup failures ("we don't have anything suitable with that name"), and treat them as compile-time errors.

The 'caller' can just say `perform a lexical lookup for the name $n$`, and the outcome is a declaration, an import prefix, or \PrependThis{}. The caller doesn't have to worry about "we don't have anything...", because we never describe propagation of compile-time errors, and the caller _will_ be able to trust that something suitable was found.

When the outcome is \PrependThis{} we know that the correct next step is to process `this.id` rather than `id`. At this point we do know that there is an instance member or an applicable extension method with the requested name; the only error that we can still encounter is that there is more than one applicable extension method.

One nice property with this approach is that it keeps _everything_ about getter/setter bundling during lookups in one location, namely here: If we do find a declaration with the right basename in an enclosing scope, but the nearest one is a getter and we requested a setter (or vice versa), we get an error. If there's nothing in any enclosing scope with that basename, but we have the same getter/setter failure with instance members, we get an error. If there is no instance member with the requested basename, but we get the same getter/setter failure with extension methods, we get an error.

So the rest of the spec doesn't need to know that getters and setters are bundled, or that basenames matter.

Returning to the two kinds of errors you mentioned:

> * We didn't find anything with the right name

Agreed that this is always an error, can be handled here.

> * We found something with the right name, but it was a setter and we wanted a getter or vice versa

This is also always an error, can be handled here.

But in both cases we have to use the right perspective: "We didn't find anything" means (1) nothing had the right basename in enclosing scopes, and (2) nothing had the right basename in the interface of `this` (if applicable), and (3) nothing had the right basename in applicable extension declarations.

"We found something with the right name, but .." means (1) the nearest enclosing scope with a declaration with the right basename had the getter/setter failure, or (2) enclosing scopes have no such basename, but the interface of `this` (if applicable) had the getter/setter failure, or (3) enclosing scopes and the interface of `this` didn't have that basename, but an applicable extension declaration had the getter/setter failure.

When none of all those errors occurred, we can say that the lexical lookup 'completes normally', and in this case we have to remember that we must prepend `this` in some cases, and we must avoid prepending `this` in other cases. In the former case we have a declaration or an import prefix; in the latter we just know that we must prepend `this` and restart the analysis.

We could of course say that we don't detect errors here, we just allow the outcome of lexical lookups to be "failed!".

But I don't think that would be helpful. For instance, if there is a top-level setter `foo=` and we are looking for a `foo` getter, we certainly shouldn't receive the outcome "failed!" from the lexical lookup, prepend `this`, and succeed in the search for an instance or extension method. We must maintain that there is an error, and it's the setter/getter failure at the top-level.

Reply:
Resolving this thread; I think the contents are covered by later threads.